src/Language/PureScript/CoreImp/Optimizer/Inliner.hs

```diff
@@ -190,37 +192,54 @@ inlineCommonOperators = everywhereTopDown $ applyAll $
     convert :: AST -> AST
     convert (App ss fn [x]) | isDict (moduleName, fnName) fn = Unary ss op x
     convert other = other
+
   mkFn :: Int -> AST -> AST
-  mkFn 0 = convert where
+  mkFn = mkFn' C.dataFunctionUncurried C.mkFn $ \ss1 ss2 ss3 args js ->
+    Function ss1 Nothing args (Block ss2 [Return ss3 js])
+
+  mkEffFn :: Int -> AST -> AST
+  mkEffFn = mkFn' C.controlMonadEffUncurried C.mkEffFn $ \ss1 ss2 ss3 args js ->
+    Function ss1 Nothing args (Block ss2 [Return ss3 (App ss3 js [])])
+
+  mkFn' :: Text -> Text -> (Maybe SourceSpan -> Maybe SourceSpan -> Maybe SourceSpan -> [Text] -> AST -> AST) -> Int -> AST -> AST
+  mkFn' modName fnName res 0 = convert where
     convert :: AST -> AST
-    convert (App _ mkFnN [Function s1 Nothing [_] (Block s2 js)]) | isNFn C.mkFn 0 mkFnN =
-      Function s1 Nothing [] (Block s2 js)
+    convert (App _ mkFnN [Function s1 Nothing [_] (Block s2 [Return s3 js])]) | isNFn modName fnName 0 mkFnN =
+      res s1 s2 s3 [] js
     convert other = other
-  mkFn n = convert where
+  mkFn' modName fnName res n = convert where
     convert :: AST -> AST
-    convert orig@(App ss mkFnN [fn]) | isNFn C.mkFn n mkFnN =
+    convert orig@(App ss mkFnN [fn]) | isNFn modName fnName n mkFnN =
       case collectArgs n [] fn of
-        Just (args, js) -> Function ss Nothing args (Block ss js)
-        Nothing -> orig
+        Just (args, [Return ss' ret]) -> res ss ss ss' args ret
+        _ -> orig
     convert other = other
     collectArgs :: Int -> [Text] -> AST -> Maybe ([Text], [AST])
     collectArgs 1 acc (Function _ Nothing [oneArg] (Block _ js)) | length acc == n - 1 = Just (reverse (oneArg : acc), js)
     collectArgs m acc (Function _ Nothing [oneArg] (Block _ [Return _ ret])) = collectArgs (m - 1) (oneArg : acc) ret
     collectArgs _ _   _ = Nothing
 
-  isNFn :: Text -> Int -> AST -> Bool
-  isNFn prefix n (Var _ name) = name == (prefix <> T.pack (show n))
-  isNFn prefix n (Indexer _ (StringLiteral _ name) (Var _ dataFunctionUncurried)) | dataFunctionUncurried == C.dataFunctionUncurried =
+  isNFn :: Text -> Text -> Int -> AST -> Bool
+  isNFn _ prefix n (Var _ name) = name == (prefix <> T.pack (show n))
```

Comment:
Yes, I defined in my `Main` module
```purescript
runFn3 f a b c = f a b c
```
and application of this was inlined, leading to runtime failure. I think this line can just be removed, the optimisation still seems to apply otherwise - I just don't understand what case this was meant to address.